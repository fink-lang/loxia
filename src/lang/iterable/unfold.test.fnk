{fink2js} = import '../../testing/generate.fnk'
{describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'



describe 'gen 2', fn:
  it 'compiles', fn:
    expect
      fink2js '
        {_unfold_} = import "@fink/js-interop/iter.js"
        unfold prev:
          prev + 1
      '
      to_match_snapshot

    expect
      fink2js '
        {_unfold_} = import "@fink/js-interop/iter.js"
        unfold prev=1:
          prev + 1
      '
      to_match_snapshot

    expect
      fink2js '
        {_unfold_} = import "@fink/js-interop/iter.js"
        unfold prev=1, acc=1:
          [prev * accu, accu + 1]
      '
      to_match_snapshot





describe 'unfold', fn:

  it 'compiles with previous', fn:
    expect
      fink2js '
        unfold prev:
          prev + 1
      '
      to_match_snapshot


  it 'compiles with default', fn:
    expect
      fink2js '
        unfold prev=1:
          prev + 1
      '
      to_match_snapshot


  it 'compiles with default and accu', fn:
    expect
      fink2js '
        unfold prev=1, acc=1:
          [prev * accu, accu + 1]
      '
      to_match_snapshot


  it 'compiles without args', fn:
    expect
      fink2js '
        unfold:
          1234
      '
      to_match_snapshot


  it 'compiles with spread', fn:
    expect
      fink2js '
        unfold:
          ...[1, 2, 3]
      '
      to_match_snapshot



