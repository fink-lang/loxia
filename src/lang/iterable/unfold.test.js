import {fink2js} from '../../testing';


test('compiles unfold', ()=> {
  expect(
    fink2js(`
      unfold curr=start:
        (curr + inc, curr * 2)

      unfold curr:
        (curr + inc, curr * 2)

      unfold:
        1234

      count = fn start=0, inc=1:
        unfold curr=start:
          curr + inc
    `)
  ).toMatchSnapshot();
});
