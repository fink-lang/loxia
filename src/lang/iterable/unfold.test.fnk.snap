// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unfold compiles destructured accus 1`] = `
"(function* unfold(ˆaccu_2) {
  while (true) {
    const [foo, ...bar] = ˆaccu_2;
    const ˆresult_1 = foo;
    yield ˆresult_1;
    ˆaccu_2 = bar;
  }
});

Object.assign(module.exports, {});"
`;

exports[`unfold compiles destructured accus 2`] = `
"(function* unfold(ˆaccu_2) {
  while (true) {
    let _do_result;

    {
      const [...ˆitems_3] = ˆaccu_2;
      _do_result = [ˆitems_3.slice(0, -1), ˆitems_3.slice(-1)];
    }
    const [foo, [bar]] = _do_result;
    _do_result = undefined;
    const ˆresult_1 = foo;
    yield ˆresult_1;
    ˆaccu_2 = bar;
  }
});

Object.assign(module.exports, {});"
`;

exports[`unfold compiles destructured accus 3`] = `
"(function* unfold(ˆaccu_2) {
  while (true) {
    const [foo, bar] = ˆaccu_2;
    const ˆresult_1 = foo;
    yield* ˆresult_1;
    ˆaccu_2 = bar;
  }
});

Object.assign(module.exports, {});"
`;

exports[`unfold compiles item is accu 1`] = `
"(function* unfold(ˆaccu_2 = 0) {
  while (true) {
    const prev = ˆaccu_2;
    const ˆresult_1 = prev + 1;
    yield ˆresult_1;
    ˆaccu_2 = ˆresult_1;
  }
});

Object.assign(module.exports, {});"
`;

exports[`unfold compiles with separate item and accu result 1`] = `
"(function* unfold(ˆaccu_2 = {}) {
  while (true) {
    const {
      \\"prev\\": prev = 0,
      \\"acc\\": acc = 0
    } = ˆaccu_2;
    const item = prev + accu;
    const ˆresult_1 = item;
    yield ˆresult_1;
    ˆaccu_2 = {
      \\"prev\\": item,
      \\"accu\\": accu + 1
    };
  }
});

Object.assign(module.exports, {});"
`;

exports[`unfold compiles with spread 1`] = `
"(function* unfold() {
  while (true) {
    const ˆresult_1 = [1, 2, 3];
    yield* ˆresult_1;
  }
});

Object.assign(module.exports, {});"
`;

exports[`unfold compiles without accus 1`] = `
"(function* unfold() {
  while (true) {
    const ˆresult_1 = 1234;
    yield ˆresult_1;
  }
});

Object.assign(module.exports, {});"
`;

exports[`unfold compiles without default assignment 1`] = `
"{
  let ˆpipe_result_1 = 0;

  ˆpipe_result_1 = function* unfold(ˆaccu_3) {
    while (true) {
      const prev = ˆaccu_3;
      const ˆresult_2 = prev + 1;
      yield ˆresult_2;
      ˆaccu_3 = ˆresult_2;
    }
  }(ˆpipe_result_1);
}
Object.assign(module.exports, {});"
`;
